proof/proof_test.go

```diff
@@ -334,12 +335,54 @@ func TestGenesisProofVerification(t *testing.T) {
 				tt, tc.assetType, tc.amount,
 				tc.tapscriptPreimage,
 			)
-			_, err := genesisProof.Verify(context.Background(), nil)
+			_, err := genesisProof.Verify(
+				context.Background(), nil, MockHeaderVerifier,
+			)
 			require.NoError(tt, err)
 		})
 	}
 }
 
+// TestProofBlockHeaderVerification ensures that an error returned by the
+// HeaderVerifier callback is correctly propagated by the Verify proof method.
+func TestProofBlockHeaderVerification(t *testing.T) {
+	t.Parallel()
+
+	proof, _ := genRandomGenesisWithProof(t, asset.Collectible, nil, nil)
+
+	// Create a base reference for the block header. We will later modify
+	// the proof block header.
+	originalBlockHeader := proof.BlockHeader
+
+	// Header verifier compares given header to expected header. Verifier
+	// does not return error.
+	errHeaderVerifier := fmt.Errorf("invalid block header")
+	headerVerifier := func(blockHeader wire.BlockHeader) error {
+		// Compare given block header against base reference block
+		// header.
+		if blockHeader != originalBlockHeader {
+			return errHeaderVerifier
+		}
+		return nil
+	}
+
+	// Verify that the original proof block header is as expected and
+	// therefore an error is not returned.
+	_, err := proof.Verify(
+		context.Background(), nil, headerVerifier,
+	)
+	require.NoError(t, err)
+
+	// Modify proof block header, then check that the verification function
+	// propagates the correct error.
+	proof.BlockHeader.Nonce += 1
+	_, actualErr := proof.Verify(
+		context.Background(), nil, headerVerifier,
+	)
+	expectedSubString := "failed to validate proof block header"
```

Comment:
I think `expectedSubString` should match the string used to build `errHeaderVerifier`? Instead of the string being used to wrap that error.

Could also do this with errors.Unwrap():

```
errVerifyInner := errors.Unwrap(actualErr)
require.Equal(t, errVerifyInner, errHeaderVerifier)
```

Reply:
Good idea!